Enable ray query for depth, segmentation and thermal cameras

srmainwaring opened the pull request:
# 🦟 Bug fix

Fixes #414

## Summary

This PR enables depth, segmentation and thermal cameras to set a ray query for the `ogre` and `ogre2` render engines.

The approach Introduces mix-in classes to provide access to underlying Ogre objects in the `OgreCamera` / `Ogre2Camera` and related camera classes.

Update the mouse handling in the `segmentation_camera` and `thermal_camera examples`.

![ogre_ray-query_thermal-camera](https://user-images.githubusercontent.com/24916364/139835717-a255b6bd-8e53-48dc-8376-9a3bc2de1b5c.gif)

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [x] Updated documentation (as needed)
- [x] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [x] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

srmainwaring commented:
Converting to draft: while this works I'm concerned that using RTTI will result in maintenance problems later.

Using a virtual function to access the `Ogre::Camera` would be better, but it's not clear to me the on best way to alter the inheritance structure or introduce a mix-in class.

iche033 commented:
looks like the problem is that the depth, thermal, segmentation camera etc does not derive from `OgreCamera` / `Ogre2Camera` so it can't access the `camera` member variable. 

Not sure what the consequence of changing them to derive from `OgreCamera` / `Ogre2Camera` but that's one possible option. 

Another workaround I can think off is to add a function function called `OgreObject()` to `OgreSensor` / `Ogre2Sensor` that returns an `Ogre::MovableObject` object, and in the derived Ogre/Ogre2 classes just return the `Ogre::Camera` object. In the ray query class, you can then call the `OgreObject()` function and cast it to a `Ogre::Camera`.

open to other ideas.

srmainwaring commented:
> Another workaround I can think off is to add a function function called `OgreObject()` to `OgreSensor` / `Ogre2Sensor` that returns an `Ogre::MovableObject` object, and in the derived Ogre/Ogre2 classes just return the `Ogre::Camera` object. In the ray query class, you can then call the `OgreObject()` function and cast it to a `Ogre::Camera`

This might be a good way forward. It retains some type safety and doesn't require a change to the inheritance structure. All the ignition cameras contain the ogre camera object required so can implement the function. We could add a parameter to request the type of object needed (similar to the `getCustomObject()` method on the `Ogre::TextureGpu` class that we added for Metal support). That allows for other objects to be requested if needed later without a further change to the interface.  I'll update the PR with the changes for review.

iche033 commented:
sounds good, thanks

srmainwaring commented:
I took a slightly different tack here: `OgreSensor` / `Ogre2Sensor` is concerned with scene nodes rather than movable objects. Rather than alter their purpose I've introduced small abstract mix-in classes to expose the underlying Ogre objects and implemented their methods in the cameras. In some cases these coincide with existing non-virtual methods on the camera class.

The mix-in class has two accessors. A specific one to access an `Ogre::Camera` and a general one to access an `Ogre::MovableObject`. I have not implemented the general case (so could remove it), but it may be useful to have if other underlying objects are required and you don't want to break ABI.

Coming up with a mix-in class name that makes sense and does not clash with existing objects is always a challenge, I went with `OgreObjectInterface` which is a bit clumsy and I'm not tied to it by any means. The camera access method has been chosen to coincide with any existing accessor already used (which happen to be different for the Ogre and Ogre2 render engines).

iche033 commented on ogre2/include/ignition/rendering/ogre2/Ogre2ObjectInterface.hh:
```diff
@@ -0,0 +1,50 @@
+    inline namespace IGNITION_RENDERING_VERSION_NAMESPACE {
+
+    /// \brief Mixin class to provide direct access to Ogre objects.
+    class Ogre2ObjectInterface
```
I think you'll need to add `IGNITION_RENDERING_OGRE2_VISIBLE` here for ignition to export the symbols, i.e.

`class IGNITION_RENDERING_OGRE2_VISIBLE Ogre2ObjectInterface`

srmainwaring replied:
Added export declarations and provided a default implementation for the `OgreMovableObject` method. Removed empty methods from the camera classes.

I'm not sure that a `const char* _typename` parameter is the best choice, something like the `Ogre::IdString` would be better but I'm not sure of the Ignition equivalent. Given that the function is a placeholder to signal intent, and that any changes to implement the method in camera classes would require a major release in any case, it's probably ok to leave it as is for now.

iche033 reviewed (commented):
ok yeah your approach is fine. Given that the `OgreMovableObject` accessor function is currently not used, one suggestion is to make it non-pure-virtual and just return `nullptr` inline so that the derived classes do not have to implement it. I think it may break ABI if the derived classes decide to override the function later but I'm leaning towards reducing the overhead of adding new camera classes for now.

iche033 commented on ogre/include/ignition/rendering/ogre/OgreThermalCamera.hh:
```diff
@@ -112,6 +117,7 @@ namespace ignition
       private: std::unique_ptr<OgreThermalCameraPrivate> dataPtr;
 
       private: friend class OgreScene;
+      private: friend class OgreRayQuery;
```
is the `friend class` still needed now that it's using `OgreObjectInterface`? Same goes for other ogre camera headers

srmainwaring replied:
You're right - they're not needed now. Originally I had the new members as protected but changed them to public to match the access level of `OgreCamera()` in `Ogre2Camera` etc. I'll remove all the friend classes that are no longer required.

iche033 reviewed (commented):
works well for me. Just one minor comment, otherwise good to go.
